group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java

```diff
@@ -157,716 +240,779 @@ public GroupAssignment build() {
+     *   <li>Member 3: 11 partitions</li>
+     * </ul>
      *
-     * The method uses a do-while loop to ensure at least one pass over the partitions and continues
-     * reassigning as long as there are modifications to the current assignments. It checks for balance
-     * after each reassignment and exits if the balance is achieved.
+     * In this assignment, the subscribers of the first topic have a difference in partitions of 1,
+     * so the topic is considered balanced. The same applies to the second topic. However, balance
+     * can be improved by moving a partition from the second topic from member 3 to member 2 and a
+     * partition from the first topic from member 2 to member 1.
      *
-     * @throws PartitionAssignorException if there are inconsistencies in expected members per partition
-     *         or if a partition is expected to already be assigned but isn't.
+     * @param topicIds          The topics to consider for reassignment. These topics must have at
+     *                          least two subscribers.
      */
-    private void performReassignments() {
-        boolean modified;
-        boolean reassignmentOccurred;
-        // Repeat reassignment until no partition can be moved to improve the balance.
-        do {
-            modified = false;
-            reassignmentOccurred = false;
-            // Reassign all reassignable partitions sorted in descending order
-            // by totalPartitions/number of subscribed members,
-            // until the full list is processed or a balance is achieved.
-            List<TopicIdPartition> reassignablePartitions = sortTopicIdPartitions(unassignedPartitions);
-
-            for (TopicIdPartition reassignablePartition : reassignablePartitions) {
-                // Only check if there is any change in balance if any moves were made.
-                if (reassignmentOccurred && isBalanced()) {
+    private void balanceTopics(Collection<Uuid> topicIds) {
+        List<Uuid> sortedTopicIds = sortTopicIds(topicIds);
+        // The index of the last topic in sortedTopicIds that was rebalanced.
+        // Used to decide when to exit early.
+        int lastRebalanceTopicIndex = -1;
+
+        MemberAssignmentBalancer memberAssignmentBalancer = new MemberAssignmentBalancer();
+
+        // An array of partitions, with partitions owned by the same member grouped together.
+        List<Integer> partitions = new ArrayList<>();
+
+        // The ranges in the partitions list assigned to members.
+        // Maintaining these ranges allows for constant time, deterministic picking of partitions
+        // owned by a given member.
+        Map<Integer, Integer> startPartitionIndices = new HashMap<>();
+        Map<Integer, Integer> endPartitionIndices = new HashMap<>(); // exclusive
+
+        // Repeat reassignment until no partition can be moved to improve the balance or we hit an
+        // iteration limit.
+        for (int i = 0; i < 16; i++) {
```

Comment:
Should we define a constant for 16? It may also be good to explain why we picked 16.

Reply:
Yes. The choice of 16 was arbitrary. I didn't want something too low (not enough time for convergence) or too high (too much cpu cost).